server/image_pull.go

```diff
@@ -80,16 +85,70 @@ func (s *Server) PullImage(ctx context.Context, req *pb.PullImageRequest) (resp
 			} else if tmpImgConfigDigest.String() == storedImage.ConfigDigest.String() {
 				log.Debugf(ctx, "image %s already in store, skipping pull", img)
 				pulled = img
+
+				// Skipped bytes metrics
+				if storedImage.Size != nil {
+					counter, err := metrics.CRIOImagePullsByNameSkipped.GetMetricWithLabelValues(img)
+					if err != nil {
+						log.Warnf(ctx, "Unable to write image pull name (skipped) metrics: %v", err)
+					} else {
+						counter.Add(float64(*storedImage.Size))
+					}
+				}
+
 				break
 			}
 			log.Debugf(ctx, "image in store has different ID, re-pulling %s", img)
 		}
 
-		_, err = s.StorageImageServer().PullImage(s.systemContext, img, &copy.Options{
-			SourceCtx:        &sourceCtx,
-			DestinationCtx:   s.systemContext,
-			OciDecryptConfig: dcc,
-		})
+		// Pull by collecting progress metrics
+		progress := make(chan types.ProgressProperties)
+		go func() {
```

Comment:
Follow-up on https://github.com/cri-o/cri-o/pull/2912#discussion_r366531015 : The core idea is using `defer { pullError <- … }` to ensure that even if the pull panics, `pullError` receives a value and the main goroutine is unblocked.

But see also https://github.com/cri-o/cri-o/pull/2912#discussion_r366792635 , removing this goroutine completely is even simpler.

Reply:
Awesome, this go routine is not needed at all. :) Thanks!